src/libxtp/gwbse/gw.cc

```diff
@@ -270,31 +270,27 @@ boost::optional<double> GW::SolveQP_Grid(double intercept0, double frequency0,
       _opt.qp_grid_spacing * double(_opt.qp_grid_steps - 1) / 2.0;
   boost::optional<double> newf = boost::none;
   double freq_prev = frequency0 - range;
-  std::pair<double, double> sigc_prev =
-      _sigma->CalcCorrelationDiagElement(gw_level, freq_prev);
-  double targ_prev = sigc_prev.first + intercept0 - freq_prev;
+  QPFunc fqp(gw_level, *_sigma.get(), intercept0);
+  std::pair<double, double> targ_prev = fqp(freq_prev);
   double qp_energy = 0.0;
   double pole_weight_max = -1.0;
   for (Index i_node = 1; i_node < _opt.qp_grid_steps; ++i_node) {
     double freq = freq_prev + _opt.qp_grid_spacing;
-    std::pair<double, double> sigc =
-        _sigma->CalcCorrelationDiagElement(gw_level, freq_prev);
-    double targ = sigc.first + intercept0 - freq;
-    if (targ_prev * targ < 0.0) {  // Sign change
-
-      QPFunc fqp(gw_level, *_sigma.get(), intercept0);
-      double f = SolveQP_Bisection(freq_prev, targ_prev, freq, targ, fqp);
-      std::pair<double, double> temp3 =
-          _sigma->CalcCorrelationDiagElement(gw_level, f);
-      double pole_weight = 1.0 / (1.0 - temp3.second);
-      if (pole_weight >= 1e-5 && pole_weight > pole_weight_max) {
+    std::pair<double, double> targ = fqp(freq);
+    if (targ_prev.first * targ.first < 0.0) {  // Sign change
+      double f =
+          SolveQP_Bisection(freq_prev, targ_prev.first, freq, targ.first, fqp);
+      std::pair<double, double> temp3 = fqp(f);
+      double pole_weight = -1.0 / temp3.second;
```

Comment:
Is this the only usage for the derivative of sigma_c? If so, it might be more efficient so simply use a first order approximation of the derivative. But then all the sigma_c derivative code is wasted...

Reply:
The newton iteration stuff still uses it, but it is also split up now into two functions.